src/client/dfuse/pil4dfs/aio.c

```diff
@@ -0,0 +1,503 @@
+	if (nr > io_depth)
+		nr = io_depth;
+
+	n_op_dfs = 0;
+	for (i = 0; i < nr; i++) {
+		if (ios[i]->aio_fildes >= FD_FILE_BASE)
+			n_op_dfs++;
+
+		op = ios[i]->aio_lio_opcode;
+		/* only support IO_CMD_PREAD and IO_CMD_PWRITE */
+		if (op != IO_CMD_PREAD && op != IO_CMD_PWRITE) {
+			DS_ERROR(EINVAL, "io_submit only supports PREAD and PWRITE for now");
+			return (-EINVAL);
+		}
+	}
+	if (n_op_dfs == 0)
+		return next_io_submit(ctx_real, nr, ios);
+
+	if (n_op_dfs != nr) {
+		DS_ERROR(EINVAL, "io_submit() does not support mixed non-dfs and dfs files yet");
+		return (-EINVAL);
+	}
+
+	aio_eq_count_local = aio_eq_count_g;
+	for (i = 0; i < nr; i++) {
+		op         = ios[i]->aio_lio_opcode;
+		fd         = ios[i]->aio_fildes - FD_FILE_BASE;
+		idx_aio_eq = (aio_ctx_obj->first_eq + i) % aio_eq_count_local;
+
+		D_ALLOC_PTR(ctx_ev);
+		if (ctx_ev == NULL)
+			return i ? i : (-ENOMEM);
+
+		rc = daos_event_init(&ctx_ev->ev, aio_eq_list[idx_aio_eq].eq, NULL);
+		if (rc) {
+			DL_ERROR(rc, "daos_event_init() failed");
+			D_GOTO(err, rc = daos_der2errno(rc));
+		}
+		d_iov_set(&iov, (void *)ios[i]->u.c.buf, ios[i]->u.c.nbytes);
+		sgl.sg_nr     = 1;
+		sgl.sg_iovs   = &iov;
+		ctx_ev->piocb = ios[i];
+		/* EQs are shared by contexts. Need to save ctx when polling EQs. */
+		ctx_ev->ctx = aio_ctx_obj;
+		if (op == IO_CMD_PREAD) {
+			rc = dfs_read(file_list[fd]->dfs_mt->dfs, file_list[fd]->file, &sgl,
+				      ios[i]->u.c.offset, &read_size, &ctx_ev->ev);
+			if (rc) {
+				rc2 = daos_event_fini(&ctx_ev->ev);
+				if (rc2)
+					DL_ERROR(rc2, "daos_event_fini() failed");
+				D_GOTO(err, rc);
+			}
+		}
+		if (op == IO_CMD_PWRITE) {
+			rc = dfs_write(file_list[fd]->dfs_mt->dfs, file_list[fd]->file, &sgl,
+				       ios[i]->u.c.offset, &ctx_ev->ev);
+			if (rc) {
+				rc2 = daos_event_fini(&ctx_ev->ev);
+				if (rc2)
+					DL_ERROR(rc2, "daos_event_fini() failed");
+				D_GOTO(err, rc);
+			}
+		}
+		atomic_fetch_add_relaxed(&aio_ctx_obj->num_op_submitted, 1);
+		atomic_fetch_add_relaxed(&aio_eq_list[idx_aio_eq].n_op_queued, 1);
+	}
+
+	return nr;
+
+err:
+	D_FREE(ctx_ev);
+
+	return i ? i : (-rc);
+}
+
+int
+io_cancel(io_context_t ctx, struct iocb *iocb, struct io_event *evt)
+{
+	d_aio_ctx_t *aio_ctx_obj = (d_aio_ctx_t *)ctx;
+	io_context_t ctx_real    = aio_ctx_obj->ctx;
+
+	if (next_io_cancel == NULL) {
+		next_io_cancel = dlsym(RTLD_NEXT, "io_cancel");
+		D_ASSERT(next_io_cancel != NULL);
+	}
+	if (!hook_enabled)
+		return next_io_cancel(ctx_real, iocb, evt);
+
+	if (iocb->aio_fildes < FD_FILE_BASE)
+		return next_io_cancel(ctx_real, iocb, evt);
+
+	/* daos_event_abort() may be used to implement this feature later. */
+	DS_ERROR(ENOSYS, "io_cancel() for DFS is not implemented");
+	return (-ENOSYS);
+}
+
+static int
+ev_enqueue(struct d_aio_ctx *ctx, struct d_aio_ev *ev)
+{
+	D_MUTEX_LOCK(&ctx->lock);
+	if (ctx->ev_queue_len >= ctx->depth) {
+		D_MUTEX_UNLOCK(&ctx->lock);
+		/* Unexpected here */
+		assert(0);
+	}
```

Comment:
You don't need to unlock on assert and this code as written will not log anything useful on failure, would this work better?
```suggestion
	D_ASSERT(ctx->depth < ctx->ev_queue_len);
```

Reply:
You are right. I will fix it as suggested. Thank you!

Reply:
Fixed as suggested. Thank you!

Reply:
This is fixed as you suggested. Thank you!